partiql-spi/src/main/kotlin/org/partiql/spi/Connector.kt

```diff
@@ -0,0 +1,48 @@
+package org.partiql.spi
+
+import org.partiql.spi.catalog.Catalog
+
+/**
+ * A Connector is responsible for instantiating [Catalog] implementations.
+ *
+ * @see Catalog
+ */
+public interface Connector {
+
+    /**
+     * Marker interface for an arbitrary context argument; marker over generics due to type erasure of generics.
+     */
+    public interface Context
+
+    /**
+     * Get (or instantiate) a [Catalog] with the given name.
+     *
+     * @param name      Catalog name
+     * @return
+     */
+    public fun getCatalog(name: String): Catalog
+
+    /**
+     * Get (or instantiate) a [Catalog] with the given name and additional context.
+     *
+     * Example:
+     *
+     * ```
+     * override fun getCatalog(context: Context): Catalog {
+     *     if (context !is MyContext) {
+     *        throw IllegalArgumentException("Unsupported context type: $context")
+     *     }
+     *     return getCatalog(context)
+     * }
+     *
+     * private fun getCatalog(context: MyContext): Catalog {
+     *      // ... get catalog from context
+     * }
+     * ```
+     *
+     * @param name      Catalog name
+     * @param context   Context is an arbitrary object
+     * @return
+     */
+    public fun getCatalog(name: String, context: Context): Catalog
```

Comment:
Given that a name is carried by the `Catalog` itself, it might not be necessary to pass it here (or not necessary to keep it in the `Catalog`). It may present the opportunity for mismatching names.

Reply:
Think of this being looking up a catalog implementation (or instantiating if it does not exist). I see this as no different than the Catalog's `getTable(name: Name): Table` which also presents the opportunity for mismatching names.

Now .. that would be a bad implementation which is kind of like returning the wrong table given the name. We can't stop a customer from implementing something incorrect.